### DISCUSSION TITLE
How to config external flash?

### DISCUSSION THREAD

**Discussion by OperatorB** [arendst/Tasmota#19010] [Q&A] [answered]
I have this [Sunon 3.5 ESP32-3248S035"](https://macsbug.wordpress.com/2022/10/02/esp32-3248s035/) device and I am wondering  how can I use the extra SPI flash storage chip added to this board.
Does Tasmota support such expansion of storage?

Here is the schematic:
![ESP32-3248S035-LCM-V1 1](https://github.com/arendst/Tasmota/assets/34603896/7f237150-4721-46bc-a7bb-2739eac34d43)
![ESP32-3248S035-MCU-V1 1](https://github.com/arendst/Tasmota/assets/34603896/e3b01ad1-129d-4815-9417-280871bf8ebc)

**Comment by Jason2866** [comment]
Configure the SD Card GPIOs in Tasmota.

**Reply by sfromis** [reply]
And better have the card formatted as FAT32

**Reply by barbudor** [reply]
How do you format a SPI Flash as FAT32 ? 🤔

**Reply by sfromis** [reply]
It was in reference to what Jason wrote about configuring the **SD Card**

And the linked page does say `SD SLOT ( Micro SD ) : CS=GPIO_5`

**Reply by barbudor** [reply]
The discussion is about using the SPI Flash

![image](https://github.com/arendst/Tasmota/assets/2996042/e00cc4e9-8ffc-4e7b-836c-135c5e143973)

**Reply by sfromis** [reply]
Well, as a SD card is both SPI and flash, terminology is not crystal clear either way.

**Reply by OperatorB** [reply]
Right, sorry for the terminology I meant the chip onboard not the SD card. I would like to use the spi flash chip U4.
So, my question came up partially due to another [comment](https://github.com/arendst/Tasmota/discussions/17939#discussioncomment-5016725) from gemu2015 which states that displays and SD card must be on HSPI (SPI 1 or 3 if you count the two inaccessible internal buses #terminology). If this is true than my board is factory messed up and should be declared mostly incompatible with tasmota32-lvgl firmware.

**Reply by sfromis** [reply]
If the only issue is whether Tasmota has support for U4, I'd not say that lack of this would mean "mostly incompatible".

**Reply by Jason2866** [reply, marked as answer]
Don't see an issue with Tasmota and this board. Just the extra SPI Flash is not useable and not needed for Tasmota.
The real question is does the used ESP32 module have PSRAM? Without LVGL is not useable.

**Reply by Jason2866** [reply]
After reading the link where the module is described. Removing U4 is the best decision. It is a design fail!

**Reply by OperatorB** [reply]
Thanks, I have removed U4 hence it is useless in this hardware config.
Regarding to PSRAM this type of board has some limited amount of PSRAM according to tasmota Information page but it is not an ESP32-WROVER-* so I am not sure what is exactly under the RF shield. The marking on top says ESP-WROOM-32. However, I am able to run tasmota32-lvgl including touch  functionality (on the capacitive version).
Lets not declare these type of boards incompatible.

**What about the SPI support? Is it true that display and SD card only supported if communicating on HSPI?**

**Reply by sfromis** [reply]
What do you specifically mean by "some limited amount of PSRAM according to tasmota Information page"? The minimum I've seen sold is 2MB, which is a fair amount for many purposes, including LVGL "within reason". Not what I'd call a "limited amount", outside of any amount arguably being a limit. Actually, it is not impossible to run a bit of LVGL without PSRAM, but it may run out of RAM with more than small setup, or if using more than quite small images.
![image](https://github.com/arendst/Tasmota/assets/47082390/7510797e-65cf-4517-8902-e8ce8582ced8)

When it comes to SPI, my understanding is that "HSPI" and "VSPI" functionally are no more than arbitrary labels for two sets of defaults gpio pins for SPI. However, the ESP32 allows any gpio to be configured for SPI, on equal terms as these defaults. What I see in the diagrams is that display and SD card are on separate SPI bus gpios, instead of sharing same pins. Tasmota does have some support for two SPI busses, but I'm not sure exactly what features supports more than the first SPI bus. I think that display can be configured for the 2nd SPI bus, and in that case it would make sense to have the SD card on the first.

Thus, I'd read the "only supported if" as using two SPI busses being a requirement to being able to use both features, which is kinda obvious as they are connected to different gpio pins already.

**Reply by OperatorB** [reply]
I thought "Free Memory" should be it but according to your snip it is not.
I have these amounts:
![image](https://github.com/arendst/Tasmota/assets/34603896/d948efea-c735-4b7d-b4e5-0c093e2ba8df)

Regarding to SPI, great news. I will do more testing as soon as my time allows. So far I was not able to make touch function work on the resistive version which uses hardware SPI pins for both display and touch (XPT2046). However, I think gemu2015 issued a pr #18480 which is now included in 13.0.0... I will test SD card functionality as well but I do not expect issue with that.

**Thank you for the quick responses!**
